### PR TITLE
Update packages and enable null safety

### PR DESCRIPTION
**Pull request by brim-borium**
This will update the packages to the latest versions using null safety and fix some codesmells

### CODE REVIEW

**Comment by brim-borium** [comment]
@itsMatoosh Currently the update to the newest versions gives me an info for some web implementations

   info • Avoid unsafe HTML APIs (assigning "src" attribute) • lib/spotify_sdk_web.dart:263:57 • unsafe_html
   info • Avoid unsafe HTML APIs (calling the 'open' method of Window) • lib/spotify_sdk_web.dart:409:23 • unsafe_html
   
   Can you have a look at that and see if this is fixable? I did a quick search, but didn't really find something to fix this 🙈

**Comment by Joran-Dob** [comment]
@brim-borium, I saw this issue appearing and searched a bit and it seems its not really fixable. As its meant really as a info prompt for the devs to be aware that they should not use user input to modify the HTML dom. 

There were a few open issues related to this warning, see: https://github.com/dart-lang/linter/issues/2348 & https://github.com/dart-lang/linter/issues/1604

I think its fine to use in the way that its being used right now, for example the method of opening a new tab/window the package is using is also used by the url_launcher package. See: https://github.com/flutter/plugins/blob/master/packages/url_launcher/url_launcher_web/lib/url_launcher_web.dart

**Review by Joran-Dob — DISMISSED** [review]
Nice, great work! 🎊

**Review by Joran-Dob — APPROVED** [review]

**Review by itsMatoosh — APPROVED** [review]
Looks all good! @brim-borium the warnings you're getting are related to how the authentication window is being opened on the web implementation. There isn't really another way to do that so I suggest ignoring the warnings.

**Review by Joran-Dob — APPROVED** [review]

**Comment by fotiDim** [comment]
Do we plan to tackle null safety in this one?

**Comment by brim-borium** [comment]
> Do we plan to tackle null safety in this one?

Yes :)

**Comment by brim-borium** [comment]
Currently we have 2 dev dependencies that do not have any preview-null-safety-package so I would suggest we update and migrate what is possible and then release a [preview version of this package](https://dart.dev/tools/pub/publishing#publishing-prereleases) in order to test this with the pub.dev version

**Comment by fotiDim** [comment]
Why is it a problem that dev dependencies do not support null safety?

**Comment by brim-borium** [comment]
> Why is it a problem that dev dependencies do not support null safety?

Not really a problem, more of an inconvenience. The migration guide (https://dart.dev/null-safety/migration-guide) states that it is better to wait for all packages to move to null safety before releasing. But as I mentioned we can just release a preview version until all packages we depend upon have been moved to null safety

**Comment by fotiDim** [comment]
If I am not mistaken this is not a blocker. The guide states `When all the dev dependencies support null safety, you can run tests with sound null safety.` Since we don't have any tests it should be fine creating a normal release with null safety enabled.

**Comment by brim-borium** [comment]
True but we still have some dependencies that are in preview 
```
logger: ^1.0.0-nullsafety.0
dio: 4.0.0-beta7
```

**Comment by fotiDim** [comment]
That is fine. We can just use their prerelease versions and have null safety support today. 
Most packages dropped their prereleases anyway and switched their main release to null safety. I don't see any reason maintaining both.

**Comment by brim-borium** [comment]
> That is fine. We can just use their prerelease versions and have null safety support today.
> Most packages dropped their prereleases anyway and switched their main release to null safety. I don't see any reason maintaining both.

Yeah sounds good :)

**Comment by brim-borium** [comment]
@fotiDim @Joran-Dob @itsMatoosh moved everything to null safety, but this can has some issues. Especially on the web implementation, because I was not always sure if intended behavior is to have a nullable object or not

**Review by Joran-Dob — APPROVED** [review]

File: lib/spotify_sdk.dart [hunk 1]
```diff
@@ -57,13 +56,13 @@ class SpotifySdk {
...
-      bool asRadio,
-      String scope,
+  static Future<bool?> connectToSpotifyRemote(
+      {required String clientId,
+      required String redirectUrl,
+      bool? asRadio,
+      String? scope,
       String playerName = 'Spotify SDK',
-      String accessToken}) async {
+      String? accessToken}) async {
```
Comment (fotiDim) [inline lib/spotify_sdk.dart:65]: I think all booleans should not be nullable and instead default to either false or true depending on what makes sense. 

`asRadio` would be false for example.
Under which cases would `connectToSpotifyRemote` return null? It should always be true or false, right?

File: lib/spotify_sdk.dart [hunk 2]
```diff
@@ -118,7 +117,7 @@ class SpotifySdk {
   /// Throws a [PlatformException] if disconnect failed
   /// Throws a [MissingPluginException] if the method is not implemented on
   /// the native platforms.
-  static Future<bool> disconnect() async {
+  static Future<bool?> disconnect() async {
```
Comment (fotiDim) [inline lib/spotify_sdk.dart:120]: Same as above

File: lib/spotify_sdk.dart [hunk 3]
```diff
@@ -544,7 +547,7 @@ class SpotifySdk {
 
   static void _logException(String method, Exception e) {
     if (e is PlatformException) {
-      var message = e.message;
+      var message = e.message!;
```
Comment (fotiDim) [inline lib/spotify_sdk.dart:550]: We can avoid the ! operator by doing the same thing you did with `e.details`. How do you call that in Dart? In Swift it is force unwrapping.

File: lib/spotify_sdk.dart (hunk 1 above)
Comment (fotiDim) [reply lib/spotify_sdk.dart:65]: Actually nil scope is valid on iOS. `Can be nil if you only need `app-remote-control`. Is it the same with Android and web?

**Comment by fotiDim** [comment]
Made a [mini PR](https://github.com/brim-borium/spotify_sdk/pull/82) on your PR

File: lib/spotify_sdk.dart (hunk 1 above)
Comment (fotiDim) [reply lib/spotify_sdk.dart]: @brim-borium any idea about this at least on Android? Is it easy for you to test it out?

Comment (brim-borium) [reply lib/spotify_sdk.dart]: WIll have a look at this tomorrow or on thursday

Comment (brim-borium) [reply lib/spotify_sdk.dart]: @fotiDim currently for android we are not using asRadio, scope or accessToken. On Android we can only set the scope when requesting the authenticationToken

Comment (brim-borium) [reply lib/spotify_sdk.dart]: `val connectionParams = ConnectionParams.Builder(clientId)
                    .setRedirectUri(redirectUrl)
                    .showAuthView(true)
                    .build()`

This is the android side, but I found [here  ](https://spotify.github.io/android-sdk/app-remote-lib/docs/index.html) that there is an additional method (getRequiredFeatures() ) to request features but no documentation on that ...
Will investigate, but in the meantime I would set asRadio default to false and leave the rest as is?

Comment (brim-borium) [reply lib/spotify_sdk.dart]: Also let the connection method return true or false instead of null

File: lib/spotify_sdk.dart [hunk 4]
```diff
@@ -94,10 +93,10 @@ class SpotifySdk {
...
   /// the native platforms.
   static Future<String> getAuthenticationToken(
-      {@required String clientId,
-      @required String redirectUrl,
-      bool asRadio,
-      String scope}) async {
+      {required String clientId,
+      required String redirectUrl,
+      bool asRadio = false,
+      String? scope}) async {
```
Comment (fotiDim) [inline lib/spotify_sdk.dart:99]: Same question as for `connectToSpotify`. Can scope be null on Android?

File: lib/spotify_sdk.dart (hunk 1 above)
Comment (fotiDim) [reply lib/spotify_sdk.dart]: Hmmm, for web scopes seem to be hardcoded and not being set from the Dart side.
```
/// spotify auth scopes
  static const List<String> authenticationScopes = [
    'streaming',
    'user-read-email',
    'user-read-private',
    'app-remote-control',
    'user-modify-playback-state',
    'playlist-read-private',
    'playlist-modify-public',
    'user-read-currently-playing'
  ];
  ```
  
  Maybe we should open an issue for this.

Comment (fotiDim) [reply lib/spotify_sdk.dart]: Regarding `getRequiredFeatures ` is it this maybe https://github.com/spotify/ios-sdk/issues/165?

Comment (brim-borium) [reply lib/spotify_sdk.dart]: > Regarding `getRequiredFeatures` is it [this](getRequiredFeatures) maybe?

Your link leads to a 404

File: lib/spotify_sdk.dart (hunk 4 above)
Comment (brim-borium) [reply lib/spotify_sdk.dart:99]: Will test this

Comment (brim-borium) [reply lib/spotify_sdk.dart:99]: Works with null, seems like there is a default scope used when this is null

File: lib/spotify_sdk.dart (hunk 1 above)
Comment (fotiDim) [reply lib/spotify_sdk.dart]: Can't find the link. It was related to getCapabilities. Nevermind

**Comment by itsMatoosh** [comment]
I made some improvements to the web null-safety in the branch "bugfix/web-player-error". I also fixed one edge-case web authorization issue on that branch. Could we maybe merge those changes into this branch since it's mostly null-safety-related improvements?

**Comment by brim-borium** [comment]
> I made some improvements to the web null-safety in the branch "bugfix/web-player-error". I also fixed one edge-case web authorization issue on that branch. Could we maybe merge those changes into this branch since it's mostly null-safety-related improvements?
> 
> Here's the PR: #85

Looks good to me :)

**Review by fotiDim — DISMISSED** [review]

**Pull request merged**